From 397257fb3e732fa99f53ea788da0a7c841d4fdca Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Thu, 11 Sep 2014 22:42:53 -0400
Subject: [infiniband] cxgb4: Ignore read reponse type 1 CQEs

Message-id: <1410475447-94392-35-git-send-email-svemuri@redhat.com>
Patchwork-id: 93823
O-Subject: [RHEL7.1 PATCH BZ 1124947 034/108] RDMA/cxgb4: Ignore read reponse type 1 CQEs
Bugzilla: 1124947
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

These are generated by HW in some error cases and need to be
silently discarded.

Signed-off-by: Steve Wise <swise@opengridcomputing.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit 70b9c66053ecadde421658b8ec808c981f2eef11)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/cxgb4/cq.c b/drivers/infiniband/hw/cxgb4/cq.c
index d6a7db2..ce468e5 100644
--- a/drivers/infiniband/hw/cxgb4/cq.c
+++ b/drivers/infiniband/hw/cxgb4/cq.c
@@ -365,8 +365,14 @@ void c4iw_flush_hw_cq(struct c4iw_cq *chp)
 
   if (CQE_OPCODE(hw_cqe) == FW_RI_READ_RESP) {
 
-   /*
-    * drop peer2peer RTR reads.
+   /* If we have reached here because of async
+    * event or other error, and have egress error
+    * then drop
+    */
+   if (CQE_TYPE(hw_cqe) == 1)
+    goto next_cqe;
+
+   /* drop peer2peer RTR reads.
     */
    if (CQE_WRID_STAG(hw_cqe) == 1)
     goto next_cqe;
@@ -511,8 +517,18 @@ static int poll_cq(struct t4_wq *wq, struct t4_cq *cq, struct t4_cqe *cqe,
   */
  if (RQ_TYPE(hw_cqe) && (CQE_OPCODE(hw_cqe) == FW_RI_READ_RESP)) {
 
-  /*
-   * If this is an unsolicited read response, then the read
+  /* If we have reached here because of async
+   * event or other error, and have egress error
+   * then drop
+   */
+  if (CQE_TYPE(hw_cqe) == 1) {
+   if (CQE_STATUS(hw_cqe))
+    t4_set_wq_in_error(wq);
+   ret = -EAGAIN;
+   goto skip_cqe;
+  }
+
+  /* If this is an unsolicited read response, then the read
    * was generated by the kernel driver as part of peer-2-peer
    * connection setup.  So ignore the completion.
    */
-- 
1.7.1